Apps/Common/EqualOrderOperators.C

```diff
@@ -0,0 +1,238 @@
+//==============================================================================
+//!
+//! \file EqualOrderOperators.C
+//!
+//! \date Jul 22 2015
+//!
+//! \author Arne Morten Kvarving / SINTEF
+//!
+//! \brief Various discrete equal-ordered operators.
+//!
+//==============================================================================
+
+#include "EqualOrderOperators.h"
+#include "FiniteElement.h"
+#include "Vec3.h"
+
+//! \brief Helper for adding an element matrix to several components.
+//! \param[out] EM The element matrix to add to.
+//! \param[in] A The scalar element matrix to add.
+//! \param[in] cmp Number of components to add matrix to
+//! \param[in] nf Number of components in total matrix.
+//! \param[in] scmp Index of first component to add matrix to.
+static void addComponents(Matrix& EM, const Matrix& A,
+                          size_t cmp, size_t nf, size_t scmp)
+{
+  if (cmp == 1 && nf == 1)
+    EM += A;
+  else
+    for (size_t i = 1; i <= A.rows(); ++i)
+      for (size_t j = 1; j <= A.cols(); ++j)
+        for (size_t k = 1; k <= cmp; ++k)
+          EM(nf*(i-1)+k+scmp,nf*(j-1)+k+scmp) += A(i, j);
+}
+
+
+//! \brief Helper applying a divergence (1) or a gradient (2) operation
+template<int Operation>
+static void DivGrad(Matrix& EM, const FiniteElement& fe,
+            double scale, int basis, int tbasis)
+{
+  size_t nsd = fe.grad(basis).cols();
+  for (size_t i = 1; i <= fe.basis(tbasis).size();++i)
+    for (size_t j = 1; j <= fe.basis(basis).size();++j)
+      for (size_t k = 1; k <= nsd; ++k) {
+        double div = fe.basis(basis)(j)*fe.grad(tbasis)(i,k)*fe.detJxW;
+        if (Operation == 2)
```

Comment:
the Operation is a template parameter so the compiler does the correct thing. it generates two copies versions of the function, picking the appropriate if path.